File: assets/codes/leetcode/src/main/java/leetcode/L_117.go
package sample

type Node struct {
	Val   int
	Left  *Node
	Right *Node
	Next  *Node
}

/*
*
解题思考
层次遍历加记忆
在层次遍历是会将同一层级的节点，存储到数组，所以一直指向的是右边
*/
func connect(root *Node) *Node {
	if root == nil {
		return nil
	}
	queue := []*Node{}
	queue = append(queue, root)
	for {
		if len(queue) == 0 {
			break
		}
		tmp := queue
		queue = []*Node{}
		for i, node := range tmp {
			if i+1 < len(tmp) {
				node.Next = tmp[i+1]
			}
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

	}
	return root
}

File: assets/codes/leetcode/src/main/java/leetcode/L_117.java
package leetcode;

import java.util.ArrayDeque;
import java.util.Queue;

public class L_117 {

    class Node {
        public int val;
        public Node left;
        public Node right;
        public Node next;

        public Node() {
        }

        public Node(int _val) {
            val = _val;
        }

        public Node(int _val, Node _left, Node _right, Node _next) {
            val = _val;
            left = _left;
            right = _right;
            next = _next;
        }
    }


    public Node connect(Node root) {
        // 层序遍历

        if (root == null) {
            return null;
        }

        Queue<Node> queue = new ArrayDeque<>();

        queue.offer(root);

        while (!queue.isEmpty()) {
            int n = queue.size();
            Node last = null;
            for (int i = 1; i <= n; i++) {
                Node f = queue.poll();
                if (f.left != null) {
                    // 先放左节点
                    queue.offer(f.left);
                }
                if (f.right != null) {
                    // 再放右节点
                    queue.offer(f.right);
                }
                if (i != 1) {
                    // 不是起始节点就执行 左边
                    last.next = f;
                }
                last = f;
            }
        }


        return root;
    }

    public static void main(String[] args) {

    }
}
